### PR TITLE
A New UnsafeNulls Language Feature for Explicit Nulls

### PR DESCRIPTION
**Pull request by noti0na1**
This PR removes the `UncheckedNull` type, and adds a new language feature `unsafeNulls` to explicit nulls, which allows unsafe null operations in a certain scope. In addition, a relaxed overriding rule is implemented for overriding Java classes. 

## Description

In the original explicit nulls design, `UncheckedNull` was introduced to allow member selections on nullable objects from Java code (with type `T | UncheckedNull`). We could define `UncheckedNull` as a type alias of `Null` or a true subtype of `Null`; however, we find it is difficult to implement in both ways. See https://github.com/lampepfl/dotty/issues/7828 for more discussion.

We decided to create a new language feature, called `unsafeNulls`. Inside this "unsafe" scope, all `T | Null` objects can be used as `T`, and `Null` keeps being a subtype of `Any`.

Users can import `scala.language.unsafeNulls` to create such scope, or use `-language:unsafeNulls` to enable this feature globally. The following unsafe null operations will apply to all nullable types:
1. select member of `T` on `T | Null` object
2. call extension methods of `T` on `T | Null`
3. convert `T1` to `T2` if `T1.stripAllNulls <:< T2.stripAllNulls` or `T1` is `Null` and `T2` has null value after erasure
4. allow equality check between `T` and `T | Null`

The intention of this `unsafeNulls` is to give users a better migration path for explicit nulls. The `UncheckedNull` is only limited to unsafe member selections. We think this approach can do more than `UncheckedNull` and in a better way.

Projects for Scala2 or regular dotty can try this by adding `-Yexplicit-nulls -language:unsafeNulls` to the compile options. A small number of manual modifications are expected (for example, some code relies on the facts of `Null <:< AnyRef`). To migrate to full explicit nulls in the future, `-language:unsafeNulls` can be dropped and add `import scala.language.unsafeNulls` only when needed.

## Examples

All the examples are under `-Yexplicit-nulls` flag.

```scala
def f(x: String): String = ???

import scala.language.unsafeNulls

val s: String | Null = ???
val a: String = s // unsafely convert String | Null to String

val b1 = s.trim() // call .trim() on String | Null unsafely
val b2 = b1.length()

f(s).trim() // pass String | Null as an argument of type String unsafely

val c: String = null // Null to String

val d1: Array[String] = ???
val d2: Array[String | Null] = d1 // unsafely convert Array[String] to Array[String | Null]
val d3: Array[String] = Array(null) // unsafe
```

Without the `unsafeNulls`, all these unsafe operations will not be compiled.

`unsafeNulls` also works for extension methods and implicit search.

```scala
import scala.language.unsafeNulls

val x = "hello, world!".split(" ").map(_.length)

given Conversion[String, Array[String]] = _ => ???

val y: String | Null = ???
val z: Array[String | Null] = y 
```

See more examples about `unsafeNulls` in `tests/explicit-nulls/pos/unsafe-*.scala`.

An example for relaxed overriding:

```java
abstract class J {
  abstract void f(String x);
}
```

```scala
// both S1 and S2 are acceptable

class S1 extends J {
  override def f(x: String) = ???
}

class S2 extends J {
  override def f(x: String | Null) = ???
}
```

See more examples about overriding in `tests/explicit-nulls/pos/override-*`.

We are also considering a Java-compatible flag, which enables unsafeNulls to all Java method calls. This will be in a seperate PR.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 1]
```diff
@@ -36,15 +37,29 @@ object ProtoTypes {
...
      *  that don't unnecessarily truncate the constraint space, returning false instead.
      */
     def necessarilyCompatible(tp: Type, pt: Type)(using Context): Boolean =
       val tpw = tp.widenExpr
       val ptw = pt.widenExpr
-      necessarySubType(tpw, ptw) || tpw.isValueSubType(ptw) || viewExists(tp, pt)
+      necessarySubType(tpw, ptw) || tpw.isValueSubType(ptw) ||
+      Nullables.convertUnsafeNulls && {
+        // See comments in `isCompatible`
+        val tpwsn = tpw.stripAllNulls
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:57]: Correction: I think it is better to just use the `UnsafeNullConversion` mode bit, or if that is not specific enough, add another mode bit that controls whether subtyping tests should ignore `Null` or  not. Basically, you can roll that in the `isSubType` tests that deal with `OrType`. I.e. You need to add
```scala
A | Null <: B   if A <: B
Null | A <: B   if A <: B
```
and for `necessarySubType`:
```scala
A <: B | Null (necessarily) if A <: B (necessarily)
A <: Null | B (necessarily) if A <: B (necessarily)
```
That way you can avoid most (maybe all?) uses of `stripNull`.

/cc @abgruszecki

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 2]
```diff
@@ -292,7 +293,10 @@ object RefChecks {
             memberTp.overrides(otherTp,
                 member.matchNullaryLoosely || other.matchNullaryLoosely ||
                 warnOnMigration(overrideErrorMsg("no longer has compatible type"),
-                   (if (member.owner == clazz) member else clazz).srcPos))
+                   (if (member.owner == clazz) member else clazz).srcPos)) ||
+            // releaxed override check for explicit nulls
+            (ctx.explicitNulls && (member.is(JavaDefined) || other.is(JavaDefined)) &&
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:298]: See my comment in ProtoTypes.scala for how to avoid `stripAllNulls` here. 

In general, it would be great if we could at least avoid the deep `stripAllNulls` everywhere because that one is really expensive.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 3]
```diff
@@ -3425,9 +3441,13 @@ class Typer extends Namer
...
         case closure(Nil, id @ Ident(nme.ANON_FUN), _)
         if defn.isFunctionType(wtp) && !defn.isFunctionType(pt) =>
-          pt match {
+          // Strip the Null type from the pt before check SAM type
+          val pt1 = if ctx.explicitNulls then pt.stripNull else pt
+          pt1 match {
             case SAMType(sam)
-            if wtp <:< sam.toFunctionType() =>
+            if wtp <:< sam.toFunctionType() ||
+              (Nullables.convertUnsafeNulls &&
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:3449]: Could be streamlined if we implement the changes to TypeComparer.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 4]
```diff
@@ -3425,9 +3441,13 @@ class Typer extends Namer
       tree match {
         case closure(Nil, id @ Ident(nme.ANON_FUN), _)
         if defn.isFunctionType(wtp) && !defn.isFunctionType(pt) =>
-          pt match {
+          // Strip the Null type from the pt before check SAM type
+          val pt1 = if ctx.explicitNulls then pt.stripNull else pt
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:3445]: This expression appears 6 times, so it's worth factoring this out into a separate method. Or maybe we can make `stripNull` return the type itself if `ctx.explicitNulls` is not set? It's worth experimenting here!

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 5]
```diff
@@ -1249,26 +1261,29 @@ class Typer extends Namer
       if (tree.tpt.isEmpty)
         meth1.tpe.widen match {
           case mt: MethodType =>
-            pt match {
+            val pt1 = if ctx.explicitNulls then pt.stripNull else pt
+            pt1 match {
               case SAMType(sam)
-              if !defn.isFunctionType(pt) && mt <:< sam =>
+              if !defn.isFunctionType(pt1) && (
+                mt <:< sam ||
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:1268]: Could be streamlined if we implement the changes to TypeComparer.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 6]
```diff
@@ -3484,26 +3504,54 @@ class Typer extends Namer
...
+          if pt.isRef(defn.AnyValClass) || pt.isRef(defn.ObjectClass) then
+            // We want to allow `null` to `AnyRef` if UnsafeNullConversion is enabled
+            if !(ctx.mode.is(Mode.UnsafeNullConversion) &&
+              treeTpe.isUnsafelyConvertable(pt)) then
+              report.error(em"the result of an implicit conversion must be more specific than $pt", tree.srcPos)
+            tree.cast(pt)
+          else
+            searchTree(tree)(failure => tryUnsafeNullConver(cannotFind(failure)))
+        else tryUnsafeNullConver(recover(NoMatchingImplicits))
+      }
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:3554]: This looks like it odes what we want. But it makes code that is already complex a lot more complicated. Are there ways to streamline this? I am not after adding more stuff, but maybe we can cleverly re-arrange the logic so it becomes clearer.

**Review by odersky — CHANGES_REQUESTED** [review]
We need a performance test for this. I tried but the scodec conflicts need to be resolved first.

Overall, I think the approach is workable, but we should work on streamlining with the goals of increasing clarity and efficiency.

**Comment by noti0na1** [comment]
@odersky I have rebased the branch and scodec. Will the performance test run if I comment "test performance please" here?

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 7]
```diff
@@ -320,29 +324,32 @@ object Implicits:
...
+      else if enableUnsafeNulls then
+        searchWithCache(eligibleCacheUnsafeNulls)
+      else
+        searchWithCache(eligibleCache)
 
-    private def computeEligible(tp: Type): List[Candidate] = /*>|>*/ trace(i"computeEligible $tp in $refs%, %", implicitsDetailed) /*<|<*/ {
+    private def computeEligible(tp: Type, enableUnsafeNulls: Boolean): List[Candidate] = /*>|>*/ trace(i"computeEligible $tp in $refs%, %", implicitsDetailed) /*<|<*/ {
       if (monitored) record(s"check eligible refs in irefCtx", refs.length)
-      val ownEligible = filterMatching(tp)
+      val c = if enableUnsafeNulls then ctx.addMode(Mode.UnsafeNullConversion) else ctx
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala:351]: The `ctx` in `computeEligible` is the `initctx` from `ImplicitRefs`, which is different from the typing context. Hence, it may not contain the `unsafeNulls` Mode.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 4 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala:3445]: If we define `stripNull` in this way, I'm worry this may cause confusion to other developers when looking at its name. How about renaming this function to `stripNullWhenExplicit`?

File: compiler/src/dotty/tools/dotc/core/CheckRealizable.scala [hunk 8]
```diff
@@ -116,14 +116,24 @@ class CheckRealizable(using Context) {
     case _: SingletonType | NoPrefix =>
       Realizable
     case tp =>
+      def checkAndType(x: TermRef, y: Type) =
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/CheckRealizable.scala]: This part is temporarily removed

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 9]
```diff
@@ -1,86 +1,88 @@
...
-    def isUncheckedNullableUnion(using Context): Boolean = {
-      val stripped = self.stripNull(true)
-      stripped ne self
+    /** Can the type has null value after erasure?
+     */
+    def isNullableAfterErasure(using Context): Boolean =  self match {
+      case tp: ClassInfo => tp.cls.isNullableClassAfterErasure
+      case tp: TypeProxy => tp.underlying.isNullableAfterErasure
+      case OrType(lhs, rhs) =>
+        lhs.isNullableAfterErasure || rhs.isNullableAfterErasure
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: Rewritten

File: compiler/src/dotty/tools/dotc/core/TypeErasure.scala [hunk 10]
```diff
@@ -646,6 +648,11 @@ class TypeErasure(isJava: Boolean, semiEraseVCs: Boolean, isConstructor: Boolean
         tpnme.WILDCARD
       case tp: WildcardType =>
         sigName(tp.optBounds)
+      case OrNull(tp) =>
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/TypeErasure.scala]: Removed

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 11]
```diff
@@ -170,8 +170,9 @@ object Types {
       case tp: ExprType => tp.resultType.isStable
       case tp: AnnotatedType => tp.parent.isStable
       case tp: AndType =>
-        tp.tp1.isStable && (realizability(tp.tp2) eq Realizable) ||
-        tp.tp2.isStable && (realizability(tp.tp1) eq Realizable)
+        def checkAndStable(x: Type, y: Type) =
+          x.isStable && ((realizability(y) eq Realizable) || y <:< x.widen)
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: This part is temporarily removed, see: tests/explicit-nulls/pos/flow-stable.scala.disabled

File: library/src/scalaShadowing/language.scala [hunk 12]
```diff
@@ -250,4 +250,6 @@ object language {
   object `3.0`
   object `3.1-migration`
   object `3.1`
+
+  object unsafeNulls
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply library/src/scalaShadowing/language.scala]: Comment added

File: compiler/src/dotty/tools/dotc/typer/Nullables.scala [hunk 13]
```diff
@@ -20,6 +20,12 @@ import ast.Trees.mods
 object Nullables:
   import ast.tpd._
 
+  /** Should we try to convert values ignoring Null type at this moment? */
+  def convertUnsafeNulls(using Context): Boolean =
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/typer/Nullables.scala]: removed

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: A special Mode `UnsafeNullsSubType` is added. When it is in the context, `TypeComparer` will consider `Null` as a subtype of all reference types.

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala (hunk 2 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/typer/RefChecks.scala]: `stripAllNulls` is no longer used

**Review by odersky — CHANGES_REQUESTED** [review]
I did a first pass. I have not looked at everything yet. The general direction looks good IMO but right now I find the logic complicated to follow because there are too many mode switches represented as boolean parameters with defaults. So before going further, I have the following suggestion:

Try to formulate the logic without using any boolean default parameters. Work on streamlining things in the context instead. Default parameters are a too easy way to hide complicated logic. Once that is done, consider bringing back some selected default parameters, but only if that leads to performance improvements or more streamlined code (but only after all other alternatives to streamline code have been considered).

**Comment by odersky** [comment]
I also added a commit with some small fixes of typos and wordings.

File: compiler/src/dotty/tools/dotc/typer/Nullables.scala [hunk 14]
```diff
@@ -20,6 +20,40 @@ import ast.Trees.mods
...
+  inline def useUnsafeNullsSubTypeIf[T](cond: Boolean)(inline op: Context ?=> T)(using Context): T =
+    val c = if cond then ctx.addMode(Mode.UnsafeNullsSubType) else ctx
+    op(using c)
+
+  /** A StickyKey for TypeApply and AppliedType in unsafe nulls,
+   *  In PostTyper, a relaxed bound check is used for these types.
+   *  Since we are unable to track `unsafeNulls` after Typer, this
+   *  key is necessary.
+   */
+  val UnsafeNullsKey = Property.StickyKey[Boolean]
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/typer/Nullables.scala]: Done, and use `hasAttachment` to check the key

File: compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala [hunk 15]
```diff
@@ -1,86 +1,70 @@
...
-          // since `stripNull((A | Null) & B)` would produce the wrong
-          // result `(A & B) | Null`.
-          val tp1s = strip(tp1)
-          val tp2s = strip(tp2)
-          if((tp1s ne tp1) && (tp2s ne tp2))
-            tp.derivedAndType(tp1s, tp2s)
-          else tp
-        case _ => tp
-      }
+    def stripNullWhenExplicit(using Context): Type = {
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala]: Renamed

File: compiler/src/dotty/tools/dotc/typer/Nullables.scala [hunk 16]
```diff
@@ -20,6 +20,40 @@ import ast.Trees.mods
...
+  /** Create a nullable type bound tree
+   *  If the lo is `Null`, `| Null` is added to hi
+   */
+  def createNullableTypeBoundsTree(lo: Tree, hi: Tree, alias: Tree = EmptyTree)(using Context): TypeBoundsTree =
+    val hiTpe = hi.typeOpt
+    val newHi = if needNullifyHi(lo.typeOpt, hiTpe) then TypeTree(OrType(hiTpe, defn.NullType, soft = false)) else hi
+    TypeBoundsTree(lo, newHi, alias)
+
+  /** Use unsafe nulls subtyping where `Null` is a subtype of all reference types */
+  inline def useUnsafeNullsSubTypeIf[T](cond: Boolean)(inline op: Context ?=> T)(using Context): T =
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/typer/Nullables.scala]: Changed to `withUnsafeNulls `

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 17]
```diff
@@ -455,28 +466,45 @@ object Types {
...
-        else NoSymbol
-      case tp: OrType =>
-        tp.join.classSymbol
-      case _: JavaArrayType =>
-        defn.ArrayClass
-      case _ =>
-        NoSymbol
-    }
+    final def classSymbol(using Context): Symbol = classSymbolWith(false)
+    final def classSymbolAfterErasure(using Context): Symbol = classSymbolWith(true)
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: Removed `classSymbolAfterErasure`

File: compiler/src/dotty/tools/dotc/ast/tpd.scala [hunk 18]
```diff
@@ -471,7 +473,7 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
 
   /** The wrapped array method name for an array of type elemtp */
   def wrapArrayMethodName(elemtp: Type)(using Context): TermName = {
-    val elemCls = elemtp.classSymbol
+    val elemCls = elemtp.classSymbolAfterErasure
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/ast/tpd.scala]: Removed `classSymbolAfterErasure`

**Comment by odersky** [comment]
Hi @noti0na1. What's the current state of this? Do you still want to work on improvements here?

I just had a look again, but I still find there are too many modes and default parameters floating around. Is there no way to bundle all of it in a mode bit in the context, say? 

Please re-assign me, when you think this is ready for review again.

File: compiler/src/dotty/tools/dotc/core/TypeApplications.scala [hunk 19]
```diff
@@ -396,43 +398,51 @@ class TypeApplications(val self: Type) extends AnyVal {
...
    *  `from` and `to` must be static classes, both with one type parameter, and the same variance.
    *  Do the same for by name types => From[T] and => To[T]
    */
-  def translateParameterized(from: ClassSymbol, to: ClassSymbol, wildcardArg: Boolean = false)(using Context): Type = self match {
+  def translateParameterized(from: ClassSymbol, to: ClassSymbol, wildcardArg: Boolean = false, withOrNull: Boolean = false)(using Context): Type = self match {
     case self @ ExprType(tp) =>
       self.derivedExprType(tp.translateParameterized(from, to))
     case _ =>
-      if (self.derivesFrom(from)) {
+      if (self.derivesFrom(from, isErased = unsafeNullsEnabled || withOrNull)) {
```
[the remark this answers is not in the document]
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/TypeApplications.scala]: Unnecessary flags and default params are removed now

File: compiler/src/dotty/tools/dotc/core/TypeApplications.scala [hunk 20]
```diff
@@ -400,44 +402,46 @@ class TypeApplications(val self: Type) extends AnyVal {
...
+  def translateParameterized(from: ClassSymbol, to: ClassSymbol, wildcardArg: Boolean = false, withOrNull: Boolean = false)(using Context): Type = self match {
     case self @ ExprType(tp) =>
       self.derivedExprType(tp.translateParameterized(from, to))
+    case OrNull(tp) =>
+      tp.translateParameterized(from, to, wildcardArg, withOrNull)
     case _ =>
-      if (self.derivesFrom(from)) {
+      if self.derivesFrom(from) then
         def elemType(tp: Type): Type = tp.widenDealias match
-          case tp: OrType =>
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/TypeApplications.scala:413]: I think it is better to remove this case, there is no need to check OrType here

File: compiler/src/dotty/tools/dotc/core/Contexts.scala [hunk 21]
```diff
@@ -632,7 +628,14 @@ object Contexts {
     def setRun(run: Run): this.type = updateStore(runLoc, run)
     def setProfiler(profiler: Profiler): this.type = updateStore(profilerLoc, profiler)
     def setNotNullInfos(notNullInfos: List[NotNullInfo]): this.type = updateStore(notNullInfosLoc, notNullInfos)
-    def setImportInfo(importInfo: ImportInfo): this.type = updateStore(importInfoLoc, importInfo)
+    def setImportInfo(importInfo: ImportInfo): this.type =
+      importInfo.mentionsFeature(nme.unsafeNulls) match
+        case Some(true) =>
+          setMode(this.mode &~ Mode.SafeNulls)
+        case Some(false) if ctx.settings.YexplicitNulls.value && !Feature.enabledBySetting(nme.unsafeNulls) =>
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/Contexts.scala:635]: TODO: Remove `!Feature.enabledBySetting(nme.unsafeNulls)`

**Comment by odersky** [comment]
@noti0na1 Can you rebase to current master please? I tried but got merge conflicts.

**Comment by noti0na1** [comment]
@odersky I have merged the current master to this branch. It is difficult to rebase at this moment since there are already many commits.

**Comment by odersky** [comment]
Can you squash the commits? To do a good job at reviewing I need to see this branch as a single commit on top of master. If you want to keep the history, maybe create a separate branch for this?

**Comment by noti0na1** [comment]
@odersky I have combined all the commits.

**Comment by odersky** [comment]
test performance please

**Review by odersky — CHANGES_REQUESTED** [review]
This PR still has a lot of complications, which might not be necessary.  To be able to merge this, it should be simplified ruthlessly. I mean not just a few cosmetic changes but every element of this PR should be put to the test whether it is really needed. I did that with #11375, and found that a lot of concepts seem not to be needed, after all. At least all the tests compile if they are removed. Among the removed constructs are:

 - The UnsafeNullsSubType mode
 - The double caches in Implicits
 - The withOrNull in translateParameterized
 - The special case in adapt
 - A few other smaller things

So, maybe it's best to take #11375 as a basis, and only add some specific changes to this simpler base that are needed for correctness.

The reason I am so strict here is that I worry about future maintainability of the compiler codebase. Maintainers need to be able to verify and modify ever single element of the codebase. Unnecessary code, and worse, unnecessary abstractions and conditions are
very bad for this since they destroy confidence that the code we deal with is there for a reason.

**Comment by odersky** [comment]
What about we try the following procedure to push this over the finish line?

 - Start with #11375 as the commonly agreed basis. (If possible, split it again in several commits that handle individual features, but this is not mandatory),
 - Look at each remaining failure in the community build and minimize it into a test. Then, add the simplest fix with a comment in the code that refers to the failing test. That way, we can have an audit trail that tells us why the new distinctions and changes are required.
  
When in doubt, let's discuss each failure and proposed fix together.

**Comment by odersky** [comment]
Turns out the only failing CB test in  #11375 is shapeless. This could be since shapeless has not been updated like it has here. @noti0na1 can you do the update a see what the status is?

**Comment by noti0na1** [comment]
@odersky Sorry for the late response. I will merge and look at these changes soon.

**Comment by noti0na1** [comment]
@odersky I have merged the changes from your PR. I think the introducing of `SafeNulls` really simplifies the complexity.

These are several tests for `eligibleCache` (searching inside and outside of unsafeNulls scope at the same time), and they all pass the test. 

I will update the document recently.

**Comment by odersky** [comment]
@noti0na1 Should we merge this then?

**Review by odersky — APPROVED** [review]
